lib/api/CDataFrameAnalysisInstrumentation.cc

```diff
@@ -64,48 +106,226 @@ void CDataFrameAnalysisInstrumentation::writer(core::CRapidJsonConcurrentLineWri
-        m_Writer->Key(PROGRESS_TAG);
-        m_Writer->Double(this->progress());
+        m_Writer->Key(MEMORY_TYPE_TAG);
+        this->writeMemory(timestamp);
+        m_Writer->Key(ANALYSIS_TYPE_TAG);
+        this->writeAnalysisStats(timestamp);
         m_Writer->EndObject();
     }
 }
 
-void CDataFrameAnalysisInstrumentation::writeMemory(std::uint32_t step) {
+std::int64_t CDataFrameAnalysisInstrumentation::memory() const {
+    return m_Memory.load();
+}
+
+void CDataFrameAnalysisInstrumentation::writeMemory(std::int64_t timestamp) {
     if (m_Writer != nullptr) {
         m_Writer->StartObject();
-        m_Writer->Key(STEP_TAG);
-        m_Writer->Uint(step);
+        m_Writer->Key(TYPE_TAG);
+        m_Writer->String(MEMORY_TYPE_TAG);
+        m_Writer->Key(JOB_ID_TAG);
+        m_Writer->String(m_JobId);
+        m_Writer->Key(TIMESTAMP_TAG);
+        m_Writer->Int64(timestamp);
         m_Writer->Key(PEAK_MEMORY_USAGE_TAG);
         m_Writer->Int64(m_Memory.load());
         m_Writer->EndObject();
     }
 }
 
+const std::string& CDataFrameAnalysisInstrumentation::jobId() const {
+    return m_JobId;
+}
+
+core::CRapidJsonConcurrentLineWriter* CDataFrameAnalysisInstrumentation::writer() {
+    return m_Writer;
+}
+
 counter_t::ECounterTypes CDataFrameOutliersInstrumentation::memoryCounterType() {
     return counter_t::E_DFOPeakMemoryUsage;
 }
 
 counter_t::ECounterTypes CDataFrameTrainBoostedTreeInstrumentation::memoryCounterType() {
     return counter_t::E_DFTPMPeakMemoryUsage;
 }
+
+void CDataFrameOutliersInstrumentation::writeAnalysisStats(std::int64_t timestamp) {
+    auto* writer{this->writer()};
+    if (writer != nullptr) {
+        writer->StartObject();
+        writer->Key(JOB_ID_TAG);
+        writer->String(this->jobId());
+        writer->Key(TIMESTAMP_TAG);
+        writer->Int64(timestamp);
+        writer->EndObject();
+    }
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::type(EStatsType type) {
+    m_Type = type;
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::iteration(std::size_t iteration) {
+    m_Iteration = iteration;
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::iterationTime(std::uint64_t delta) {
+    m_IterationTime = delta;
+    m_ElapsedTime += delta;
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::lossType(const std::string& lossType) {
+    m_LossType = lossType;
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::lossValues(std::string fold,
+                                                           TDoubleVec&& lossValues) {
+    m_LossValues.emplace(std::make_pair(fold, lossValues));
+}
+
+void CDataFrameTrainBoostedTreeInstrumentation::writeAnalysisStats(std::int64_t timestamp) {
+    auto* writer{this->writer()};
+    if (writer != nullptr) {
+        writer->StartObject();
+        writer->Key(JOB_ID_TAG);
+        writer->String(this->jobId());
+        writer->Key(TIMESTAMP_TAG);
+        writer->Int64(timestamp);
+        if (m_Type == E_Regression) {
```

Comment:
Maybe migrate to switch statement: if we ever add more types the missing cases would then generate a warning.